### ISSUE TITLE
Exponents not allowed in real literals

### ISSUE THREAD

**Issue by atomb** [dafny-lang/dafny#1868]
The Dafny parser seems to expect that the letter `e` may appear in decimal strings to support exponential notation, as seen here:

https://github.com/dafny-lang/dafny/blob/c10abd8670e4a4269ba388cc8e4ed1d9a103b6ad/Source/Dafny/Dafny.atg#L4759-L4777

However, this support seems incomplete or incorrect. For example, the following Dafny program is accepted, as-is, but fails with the message "rbrace expected" if `2.0` is replaced by `2e0` or `2.0e0`.

````
function use_exp(): real { 2.0 }
````

**Comment by cpitclaudel** [comment]
I think that code is trying to *reject* cases with an `e` in the string, but I don't see why that's necessary: the rule `decimaldigits` does not allow anything except numbers and `_.`, so `x.val.IndexOf('e')` should always be false.  Let's try removing that code entirely?

**Comment by atomb** [comment]
Yeah, I think you're right, and that code likely should be removed or clarified.

But it seems to be operating on the assumption that they could reasonably appear in numeric literals, suggesting that that was the intended design of Dafny. I came across this because I was trying to fill in the TODO in the reference manual about describing the syntax of real literals with exponents. And it seems that they aren't actually supported, even though they arguably should be (and are supported by Boogie), and perhaps were intended to be.

**Comment by cpitclaudel** [comment]
Good point, but if in 8 years since reals were introduced no one has needed exponents then we probably don't need to worry too much about them right now ^^

**Comment by cpitclaudel** [comment]
@parno do you remember why you introduced this extra check? f335732e

**Comment by atomb** [comment]
> Good point, but if in 8 years since reals were introduced no one has needed exponents then we probably don't need to worry too much about them right now ^^

Yeah, perhaps so. But I think it's worth having an issue even if we don't get to implementing it anytime soon.

**Comment by parno** [comment]
> Do you remember why you introduced this extra check?

No, at this point it looks mysterious (and redundant) to me as well.  Sorry!
